### PR TITLE
Add keysmith

### PR DESCRIPTION
**Pull request by SCOTT-HAMILTON**
###### Motivation for this change
Needed a simple otp client, while otpgen is still in stand-by.

###### Things done
Successfully build, tested and added to all-packages.nix

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by wamserma** [comment]
Your name is still missing from the maintainers list.

**Comment by SCOTT-HAMILTON** [comment]
Fixed

File: pkgs/tools/security/keysmith/default.nix [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    rev = "v${version}";
+    sha256 = "15fzf0bvarivm32zqa5w71mscpxdac64ykiawc5hx6kplz93bsgx";
+  };
+
+  nativeBuildInputs = [ cmake extra-cmake-modules ];
+
+  buildInputs = [ oathToolkit kirigami2 qtquickcontrols2 qtbase ];
+
+  meta = {
+    description = "OTP client for Plasma Mobile and Desktop.";
```
Comment (IvarWithoutBones) [inline pkgs/tools/security/keysmith/default.nix:28]: This shouldn't have a period at the end, as the [contributing guidelines](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md) state.

File: pkgs/tools/security/keysmith/default.nix [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    repo = "keysmith";
+    rev = "v${version}";
+    sha256 = "15fzf0bvarivm32zqa5w71mscpxdac64ykiawc5hx6kplz93bsgx";
+  };
+
+  nativeBuildInputs = [ cmake extra-cmake-modules ];
+
+  buildInputs = [ oathToolkit kirigami2 qtquickcontrols2 qtbase ];
+
+  meta = {
```
Comment (IvarWithoutBones) [inline pkgs/tools/security/keysmith/default.nix:27]: The usage of `stdenv.lib` and `lib` is inconsistent.
```suggestion
  meta = with stdenv.lib; {
    description = "OTP client for Plasma Mobile and Desktop";
    license = licenses.gpl3;
    homepage = "https://github.com/KDE/keysmith";
    maintainers = with maintainers; [ shamilton ];
    platforms = platforms.linux;
  };
```
This way you can also lose the `lib` attribute at the top.

**Review by IvarWithoutBones — CHANGES_REQUESTED** [review]

**Comment by IvarWithoutBones** [comment]
https://github.com/NixOS/nixpkgs/pull/88614/commits/6c70ff21d40c9a45a4c4ff55abb3607e928dcd34 and https://github.com/NixOS/nixpkgs/pull/88614/commits/2a224460c0489768eb42b6077359ab753b7d6016 should be squashed. 

https://github.com/NixOS/nixpkgs/pull/88614/commits/9906381129328d2e3353f18605b289665ba420bc has nothing to do with keysmith itself, should lose that in the commit message.

File: pkgs/tools/security/keysmith/default.nix [hunk 3]
```diff
@@ -0,0 +1,33 @@
+{ stdenv
+, fetchFromGitHub
+, cmake
+, extra-cmake-modules
+, qtbase
+, qtquickcontrols2
+, kirigami2
+, oathToolkit
+}:
```
Comment (jonringer) [inline pkgs/tools/security/keysmith/default.nix:9]:
```suggestion
}:

```

File: pkgs/tools/security/keysmith/default.nix [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+stdenv.mkDerivation rec {
+
```
Comment (jonringer) [inline pkgs/tools/security/keysmith/default.nix:11]:
```suggestion
```

File: pkgs/tools/security/keysmith/default.nix [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  pname = "Keysmith";
```
Comment (jonringer) [inline pkgs/tools/security/keysmith/default.nix:12]: generally all lower case
```suggestion
  pname = "keysmith";
```

File: pkgs/tools/security/keysmith/default.nix [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+  nativeBuildInputs = [ cmake extra-cmake-modules ];
+
+  buildInputs = [ oathToolkit kirigami2 qtquickcontrols2 qtbase ];
+
+  meta = with stdenv.lib; {
+    description = "OTP client for Plasma Mobile and Desktop";
+    license = licenses.gpl3;
+    homepage = "https://github.com/KDE/keysmith";
+    maintainers = [ "Scott Hamilton <sgn.hamilton+nixpkgs@protonmail.com>" ];
```
Comment (jonringer) [inline pkgs/tools/security/keysmith/default.nix:30]: should refer to your maintainer entry
```suggestion
    maintainers = with maintainers; [ shamilton ];
```

File: pkgs/tools/security/keysmith/default.nix [hunk 7]
```diff
@@ -0,0 +1,33 @@
+{ stdenv
+, fetchFromGitHub
+, cmake
+, extra-cmake-modules
+, qtbase
+, qtquickcontrols2
+, kirigami2
+, oathToolkit
+}:
+stdenv.mkDerivation rec {
```
Comment (jonringer) [inline pkgs/tools/security/keysmith/default.nix:10]: please read https://nixos.org/nixpkgs/manual/#sec-language-qt
```suggestion
mkDerivation rec {
```

File: pkgs/tools/security/keysmith/default.nix [hunk 8]
```diff
@@ -0,0 +1,33 @@
+{ stdenv
```
Comment (jonringer) [inline pkgs/tools/security/keysmith/default.nix:1]: https://nixos.org/nixpkgs/manual/#sec-language-qt
```suggestion
{ lib
, mkDerivation
```

File: pkgs/tools/security/keysmith/default.nix [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    repo = "keysmith";
+    rev = "v${version}";
+    sha256 = "15fzf0bvarivm32zqa5w71mscpxdac64ykiawc5hx6kplz93bsgx";
+  };
+
+  nativeBuildInputs = [ cmake extra-cmake-modules ];
+
+  buildInputs = [ oathToolkit kirigami2 qtquickcontrols2 qtbase ];
+
+  meta = with stdenv.lib; {
```
Comment (jonringer) [inline pkgs/tools/security/keysmith/default.nix:26]:
```suggestion
  meta = with lib; {
```

**Review by jonringer — COMMENTED** [review]
please cleanup git history as well, should only have 2 commits:
```
maintainers: add shamilton
keysmith: init at 0.1
```

**Comment by jonringer** [comment]
Also, thanks for opening your first PR with nixpkgs :)

**Comment by SCOTT-HAMILTON** [comment]
> Also, thanks for opening your first PR with nixpkgs :)

You're welcome :+1: (Even though it's my third)

**Comment by SCOTT-HAMILTON** [comment]
Rebased to new branch, moved to https://github.com/NixOS/nixpkgs/pull/89413

**Comment by SCOTT-HAMILTON** [comment]
Discovered git force pushing so coming back to this thread and closing the new one.

**Review by jonringer — COMMENTED** [review]
I get issues trying to launch it:
```
[nix-shell:/home/jon/.cache/nixpkgs-review/pr-88614]$ ./results/keysmith/bin/org.kde.keysmith
QQmlApplicationEngine failed to load component
qrc:/main.qml:29 Type Kirigami.ApplicationWindow unavailable
file:///nix/store/8jhqa8q59r448c94yj88w2f1s1l4cmb8-kirigami2-5.68.0/lib/qt-5.12.7/qml/org/kde/kirigami.2/styles/org.kde.desktop/ApplicationWindow.qml:12 Type Base.ApplicationWindow unavailable
file:///nix/store/8jhqa8q59r448c94yj88w2f1s1l4cmb8-kirigami2-5.68.0/lib/qt-5.12.7/qml/org/kde/kirigami.2/ApplicationWindow.qml:10 module "QtGraphicalEffects" is not installed
```
also, do you mind adding a link:
```
  postInstall = ''
     ln -s $out/bin/org.kde.keysmith $out/bin/keysmith
  '';
```

**Comment by SCOTT-HAMILTON** [comment]
I've wrapped the needed QT and QML ENV vars in postInstall, I think this should fix your issue (even though I couldn't reproduce it).

**Comment by jonringer** [comment]
> I've wrapped the needed QT and QML ENV vars in postInstall, I think this should fix your issue (even though I couldn't reproduce it).

I don't run a DE, so I do usually have issues with quite a few KDE / Qt / GTK apps

**Review by jonringer — APPROVED** [review]
diff LGTM
gui launches fine

```
[2 built, 1 copied (0.5 MiB), 0.1 MiB DL]
https://github.com/NixOS/nixpkgs/pull/88614
1 package built:
keysmith
```

**Pull request merged**